File: spikes/04-bench.mjs
// Spike 4 (JS side): same invert flow through the prototype, median of 10.
// Also isolates where time goes (unzip / mutate / rezip) and what deflate
// level costs, since zip write dominated spike 1.
import { readFile } from "node:fs/promises";
import { deflateSync } from "fflate";
import { Presentation } from "./lib/pptx.mjs";
import { readZip, writeZip } from "./lib/opc.mjs";

const fixture = process.argv[2] ?? "../tests/fixtures/hagar-presentatie.pptx";
const bytes = new Uint8Array(await readFile(fixture));

function invertOnce() {
  const pres = Presentation.open(bytes);
  for (const slide of pres.slides) {
    slide.setBackground("1A1A2E");
    for (const shape of slide.shapes()) {
      if (shape.hasTextFrame) for (const run of shape.runs()) run.setFontColor("E8E8F0");
    }
  }
  return pres.save().length;
}

invertOnce(); // warmup
const times = [];
for (let i = 0; i < 10; i++) {
  const t0 = performance.now();
  invertOnce();
  times.push(performance.now() - t0);
}
times.sort((a, b) => a - b);
console.log(`prototype invert+save median: ${times[5].toFixed(1)}ms (min ${times[0].toFixed(1)}ms)`);

// --- phase breakdown -------------------------------------------------------
let t = performance.now();
const parts = readZip(bytes);
const unzipMs = performance.now() - t;
t = performance.now();
writeZip(parts);
const rezipMs = performance.now() - t;
console.log(`phases: unzip ${unzipMs.toFixed(1)}ms, rezip ${rezipMs.toFixed(1)}ms (rest = parse+mutate)`);

// --- deflate level cost on the biggest part -------------------------------
const biggest = [...parts.values()].sort((a, b) => b.data.length - a.data.length)[0].data;
for (const level of [1, 6, 9]) {
  const t0 = performance.now();
  const out = deflateSync(biggest, { level });
  console.log(`deflate level ${level}: ${(performance.now() - t0).toFixed(1)}ms -> ${(out.length / 1024).toFixed(0)} KiB (input ${(biggest.length / 1024).toFixed(0)} KiB)`);
}

File: spikes/04-bench.py
"""Spike 4 (Python side): time the same invert flow in python-pptx."""
import io
import sys
import time

from pptx import Presentation
from pptx.dml.color import RGBColor

path = sys.argv[1] if len(sys.argv) > 1 else "../tests/fixtures/hagar-presentatie.pptx"
data = open(path, "rb").read()

BG = RGBColor(0x1A, 0x1A, 0x2E)
FG = RGBColor(0xE8, 0xE8, 0xF0)


def run_once() -> int:
    prs = Presentation(io.BytesIO(data))
    for slide in prs.slides:
        fill = slide.background.fill
        fill.solid()
        fill.fore_color.rgb = BG
        for shape in slide.shapes:
            if shape.has_text_frame:
                for para in shape.text_frame.paragraphs:
                    for run in para.runs:
                        run.font.color.rgb = FG
    buf = io.BytesIO()
    prs.save(buf)
    return buf.tell()


run_once()  # warmup
times = []
for _ in range(10):
    t0 = time.perf_counter()
    run_once()
    times.append((time.perf_counter() - t0) * 1000)
times.sort()
print(f"python-pptx invert+save median: {times[len(times) // 2]:.1f}ms (min {times[0]:.1f}ms)")
